Admission control to prevent exec on privileged pods

csrwng opened the pull request:
Add admission control to the CONNECT action in the API server and create an admission controller that will deny pod/exec to pods with privileged containers.

csrwng committed:
Add admission control to the Connect method in the API Server

The resource passed to admission control is a ConnectRequest object
which includes additional information about the current request.

csrwng committed:
Admission control to prevent exec on privileged pods

csrwng commented:
@smarterclayton @derekwaynecarr ptal
@deads2k fyi

derekwaynecarr commented:
Is there a reason we would not want to enable this plugin by default on all providers?

Supporting privileged containers in general is a flag in the salt configuration.  

Sent from my iPhone

> On May 13, 2015, at 10:45 PM, Cesar Wong <notifications@github.com> wrote:
> 
> Add admission control to the CONNECT action in the API server and create an admission controller that will deny pod/exec to pods with privileged containers.
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/GoogleCloudPlatform/kubernetes/pull/8233
> 
> Commit Summary
> 
> Use resource path for API server request context
> Add admission control to the Connect method in the API Server
> Admission control to prevent exec on privileged pods
> File Changes
> 
> M cmd/kube-apiserver/app/plugins.go (1)
> M pkg/api/rest/types.go (9)
> M pkg/apiserver/api_installer.go (4)
> M pkg/apiserver/resthandler.go (11)
> A plugin/pkg/admission/exec/denyprivileged/admission.go (81)
> A plugin/pkg/admission/exec/denyprivileged/admission_test.go (90)
> Patch Links:
> 
> https://github.com/GoogleCloudPlatform/kubernetes/pull/8233.patch
> https://github.com/GoogleCloudPlatform/kubernetes/pull/8233.diff
> —
> Reply to this email directly or view it on GitHub.
>

csrwng commented:
@derekwaynecarr no reason

csrwng committed:
Add operation checking to admission control handlers

Adds a new method to the handler interface that returns true only if the
admission control handler handles that operation.

derekwaynecarr commented:
@csrwng can you poke when ready for final review?

csrwng commented:
@derekwaynecarr - should be ready for another review

derekwaynecarr commented on pkg/admission/chain.go:
```diff
@@ -39,6 +39,9 @@ func NewFromPlugins(client client.Interface, pluginNames []string, configFilePat
 // Admit performs an admission control check using a chain of handlers, and returns immediately on first error
 func (admissionHandler chainAdmissionHandler) Admit(a Attributes) error {
 	for _, handler := range admissionHandler {
+		if !handler.Handles(a.GetOperation()) {
```
I know we don't have one now, but we should have one.  Please add a test for the chainAdmissionHandler that takes a set of mock admission handlers, and ensure that it respects the handles method for the objects it delegates.

derekwaynecarr commented on pkg/admission/chain.go:
```diff
@@ -46,3 +49,13 @@ func (admissionHandler chainAdmissionHandler) Admit(a Attributes) error {
 }
+
+// Handles will return true if any of the handlers handles the given operation
+func (admissionHandler chainAdmissionHandler) Handles(operation string) bool {
```
Needs a test case.

derekwaynecarr commented on pkg/admission/interfaces.go:
```diff
@@ -36,4 +36,8 @@ type Attributes interface {
 	Admit(a Attributes) (err error)
+
+	// Handles returns true if this admission controller can handle the given operation
+	// where operation can be one of CREATE, UPDATE, DELETE, or CONNECT
```
Add constants for the enumerated set of options so its clear in the long term what operations are supported in Kubernetes.  I think it's ok for it to remain a string as part of this PR, I just want a constant to find all future references, and this PR seems like a good one to do it.

derekwaynecarr commented on pkg/api/rest/types.go:
```diff
@@ -87,3 +87,13 @@ func (svcStrategy) AllowCreateOnUpdate() bool {
+// ConnectRequest is an object passed to admission control for Connect operations
+type ConnectRequest struct {
+	Name     string
+	Options  runtime.Object
```
This is not obvious to me right now, it probably will be later as I review what Options is.

derekwaynecarr commented on pkg/api/rest/types.go:
```diff
@@ -87,3 +87,13 @@ func (svcStrategy) AllowCreateOnUpdate() bool {
+type ConnectRequest struct {
+	Name     string
+	Options  runtime.Object
+	RESTPath string
```
Is this path plus subpath?

derekwaynecarr replied:
ah, ok, so this is path the storage was registered to work against.

```
pods/exec
```

derekwaynecarr commented on plugin/pkg/admission/exec/denyprivileged/admission.go:
```diff
@@ -0,0 +1,81 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
2015

I think the boilerplate stuff should have caught this.

derekwaynecarr commented on plugin/pkg/admission/exec/denyprivileged/admission_test.go:
```diff
@@ -0,0 +1,90 @@
+/*
+Copyright 2014 The Kubernetes Authors All rights reserved.
```
2015

derekwaynecarr commented on plugin/pkg/admission/namespace/autoprovision/admission_test.go:
```diff
@@ -81,30 +81,6 @@ func TestAdmissionNamespaceExists(t *testing.T) {
 }
 
-// TestIgnoreAdmission validates that a request is ignored if its not a create
-func TestIgnoreAdmission(t *testing.T) {
```
Ideally, we would have a test that wraps the handler in chainAdmissionHandler and ensures that the request was ignored.  Also it would be good to have tests that Handler returns the expected values since a bad value could cause null pointers.  This probably applies to all the plug-ins, but bugs me most on tests where we were testing that we ignore.

derekwaynecarr commented on plugin/pkg/admission/namespace/lifecycle/admission_test.go:
```diff
@@ -64,17 +64,4 @@ func TestAdmission(t *testing.T) {
 		t.Errorf("Expected error rejecting creates in a namespace when it is terminating")
 	}
-
-	// verify update operations in the namespace can proceed
```
maybe wrap our handler with the chainAdmissionHandler and keep this test block?

derekwaynecarr commented on plugin/pkg/admission/securitycontext/scdeny/admission.go:
```diff
@@ -34,20 +34,21 @@ func init() {
 func NewSecurityContextDeny(client client.Interface) admission.Interface {
-	return &plugin{client}
+	return &plugin{
+		Handler: admission.NewHandler("CREATE", "UPDATE"),
```
add a test to admission_test.go to validate what the expected Handlers are.

derekwaynecarr commented:
Structure looks ok, I would like to make your PR the PR that makes people code against constants for operation names.  Let me know when updates are made and will do another pass.

csrwng commented:
Code updated. Should be ready for another pass. Thx for the review.

derekwaynecarr commented:
Did a quick pass and things look good.  Once merge freeze is unfrozen, will take a final pass. 

Sent from my iPhone

> On May 20, 2015, at 5:31 PM, Cesar Wong <notifications@github.com> wrote:
> 
> Code updated. Should be ready for another pass. Thx for the review.
> 
> —
> Reply to this email directly or view it on GitHub.
>

csrwng commented:
Fixed race condition in unit tests

derekwaynecarr commented on pkg/admission/interfaces.go:
```diff
@@ -36,4 +36,19 @@ type Attributes interface {
+
+// Operation constants
+const (
+	CREATE  Operation = "CREATE"
```
Effective Go style conventions, and the standard library all appear to use MixedCase for constants.

please update to just admission.Create, admission.Update, admission.Delete, admission.Connect

derekwaynecarr commented:
Fix the ALLCAPS const, then LGTM.

csrwng commented:
Constants renamed

derekwaynecarr commented:
LGTM, merging.
